lib/spack/spack/test/cmd/env.py

```diff
@@ -3031,6 +3031,34 @@ def test_stack_view_activate_from_default(
         assert "FOOBAR=mpileaks" in shell
 
 
+def test_envvar_set_in_activate(tmpdir, mock_fetch, mock_packages, mock_archive, install_mockery):
+    filename = str(tmpdir.join("spack.yaml"))
+    with open(filename, "w", encoding="utf-8") as f:
+        f.write(
+            """\
+spack:
+  specs:
+    - cmake%gcc
+  env_vars:
+    set:
+      ENVAR_SET_IN_ENV_LOAD: "True"
+"""
+        )
+    with tmpdir.as_cwd():
+        env("create", "test", "./spack.yaml")
+        with ev.read("test"):
+            install()
+
+        test_env = ev.read("test")
+
+        output = env("activate", "--sh", "test")
+        assert "ENVAR_SET_IN_ENV_LOAD=True" in output
+
+        with test_env:
+            output = env("deactivate", "--sh")
+            assert "ENVAR_SET_IN_ENV_LOAD" not in output
```

Comment:
Ok one more thing here. I'm forgetting that activating an env with `with env:` on the python side doesn't actually set the env vars -- it just makes that environment the active environment. That's intentional in that we don't want to screw up the Spack runtime environment with things like PYTHONPATH, etc. from activating various environments.

In addition to that, the various `EnvironmentModifications` objects are dynamically checking whether things are in the environment and adding code to the shell output only if it's needed. So this is why your test succeeds -- the env var you want to unset is not set in the first place.

What you need to do here is this:

```suggestion
        with test_env:
            with spack.util.environment.set_env(ENVAR_SET_IN_ENV_LOAD=True):
                output = env("deactivate", "--sh")
                assert "unset ENVAR_SET_IN_ENV_LOAD" in output
```

Reply:
I that change and the test passes and both asserts look like they are looking in valid output

for  `assert "ENVAR_SET_IN_ENV_LOAD=True" in output`, `output` is:
```
export SPACK_ENV=/tmp/pytest-of-jfrye/pytest-38/test_envvar_set_in_activate_mo0/mock-env-path/test;
export SPACK_ENV_VIEW=default;
alias despacktivate='spack env deactivate';
export ACLOCAL_PATH=/tmp/pytest-of-jfrye/pytest-38/test_envvar_set_in_activate_mo0/mock-env-path/test/.spack-env/view/share/aclocal;
export CMAKE_PREFIX_PATH=/tmp/pytest-of-jfrye/pytest-38/test_envvar_set_in_activate_mo0/mock-env-path/test/.spack-env/view;
export ENVAR_SET_IN_ENV_LOAD=True;
export MANPATH=/tmp/pytest-of-jfrye/pytest-38/test_envvar_set_in_activate_mo0/mock-env-path/test/.spack-env/view/share/man:/tmp/pytest-of-jfrye/pytest-38/test_envvar_set_in_activate_mo0/mock-env-path/test/.spack-env/view/man:;
export PATH=/tmp/pytest-of-jfrye/pytest-38/test_envvar_set_in_activate_mo0/mock-env-path/test/.spack-env/view/bin:/home/jfrye/.spack/bootstrap/environments/python3.6-x86_64-7d8fd/view/bin:/home/jfrye/.spack/bootstrap/store/linux-centos7-x86_64/gcc-10.2.1/patchelf-0.17.2-o6soxsz4hwdhzzbu4j56zwcclqhram25/bin:/home/jfrye/test/spack/bin:/usr/local/bin:/usr/bin:/usr/local/sbin:/usr/sbin;
export PKG_CONFIG_PATH=/tmp/pytest-of-jfrye/pytest-38/test_envvar_set_in_activate_mo0/mock-env-path/test/.spack-env/view/share/pkgconfig:/tmp/pytest-of-jfrye/pytest-38/test_envvar_set_in_activate_mo0/mock-env-path/test/.spack-env/view/lib64/pkgconfig:/tmp/pytest-of-jfrye/pytest-38/test_envvar_set_in_activate_mo0/mock-env-path/test/.spack-env/view/lib/pkgconfig;
```
and for `assert "unset ENVAR_SET_IN_ENV_LOAD" in output`, `output` is:
```
if [ ! -z ${SPACK_ENV+x} ]; then
unset SPACK_ENV; export SPACK_ENV;
fi;
if [ ! -z ${SPACK_ENV_VIEW+x} ]; then
unset SPACK_ENV_VIEW; export SPACK_ENV_VIEW;
fi;
alias despacktivate > /dev/null 2>&1 && unalias despacktivate;
if [ ! -z ${SPACK_OLD_PS1+x} ]; then
    if [ "$SPACK_OLD_PS1" = '$$$$' ]; then
        unset PS1; export PS1;
    else
        export PS1="$SPACK_OLD_PS1";
    fi;
    unset SPACK_OLD_PS1; export SPACK_OLD_PS1;
fi;
unset ENVAR_SET_IN_ENV_LOAD;
```

Reply:
@tgamblin I think its testing correctly. Can you take a look when you get a chance?